### PR TITLE
Registry logic restructuring

### PR DESCRIPTION
**Pull request by dbw7**
This PR contains a lot of restructuring based on the conversations we had. A lot of the tests need to be updated and this implementation is not complete but I’m putting it up as a draft to get a second set of eyes on it. There also may be some lingering parts that don’t necessarily make sense at the moment.

To this though, I’ve tested the current implementation and this is deploying exactly what was being deployed before the restructure without any change to the image definition file.

### CODE REVIEW

File: pkg/image/definition.go [hunk 1]
```diff
@@ -120,10 +120,11 @@ type HelmChart struct {
...
 type Kubernetes struct {
-	Version   string    `yaml:"version"`
-	Network   Network   `yaml:"network"`
-	Nodes     []Node    `yaml:"nodes"`
-	Manifests Manifests `yaml:"manifests"`
+	Version    string      `yaml:"version"`
+	Network    Network     `yaml:"network"`
+	Nodes      []Node      `yaml:"nodes"`
+	Manifests  Manifests   `yaml:"manifests"`
+	HelmCharts []HelmChart `yaml:"charts"`
```
Comment (atanasdinov) [inline pkg/image/definition.go:127]: This is not yet dropped from the `EmbeddedArtifactRegistry`.

File: pkg/combustion/templates/13-embedded-registry.sh.tpl [hunk 2]
```diff
@@ -4,7 +4,7 @@ set -euo pipefail
 mkdir /opt/hauler
 
 mount /usr/local
-mv {{ .EmbeddedRegistryTar }} /opt/hauler/{{ .EmbeddedRegistryTar }}
+mv ./registry/* /opt/hauler/
```
Comment (atanasdinov) [inline pkg/combustion/templates/13-embedded-registry.sh.tpl:7]: I'd suggest to template the directory value. As mentioned in one of the previous PRs, it's very easy for the Go implementation to start using something different than a `registry` directory which will completely invalidate the script but the tests would continue passing.

File: pkg/combustion/registry.go [hunk 3]
```diff
@@ -173,6 +169,28 @@ func createRegistryCommand(ctx *image.Context, commandName string, args []string
 	return cmd, logFile, nil
 }
 
-func IsEmbeddedArtifactRegistryEmpty(registry image.EmbeddedArtifactRegistry) bool {
-	return len(registry.HelmCharts) == 0 && len(registry.ContainerImages) == 0
+func IsEmbeddedArtifactRegistryAndKubernetesManifestsEmpty(ctx *image.Context) bool {
```
Comment (atanasdinov) [inline pkg/combustion/registry.go:172]: This is now too complex. I'd suggest to invert the naming and usage.
```suggestion
func IsEmbeddedArtifactRegistryConfigured(ctx *image.Context) bool {
```

File: pkg/combustion/registry.go [hunk 4]
```diff
@@ -30,50 +28,48 @@ var haulerManifest string
...
+	haulerBinaryPath := fmt.Sprintf("hauler-%s", string(ctx.ImageDefinition.Image.Arch))
+	err := copyHaulerBinary(ctx, haulerBinaryPath)
 	if err != nil {
 		log.AuditComponentFailed(registryComponentName)
-		return nil, fmt.Errorf("populating hauler store: %w", err)
+		return nil, fmt.Errorf("copying hauler binary: %w", err)
 	}
 
-	err = generateRegistryTar(ctx)
+	registryScriptNameResult, err := writeRegistryScript(ctx)
```
Comment (atanasdinov) [inline pkg/combustion/registry.go:43]: This is the operation which should be the last in this component. All other steps are the prerequisites of having the final Combustion script written to the filesystem. This would also make templating the "registry" directory easier.

File: pkg/combustion/registry.go [hunk 5]
```diff
@@ -30,50 +28,48 @@ var haulerManifest string
...
 	if err != nil {
 		log.AuditComponentFailed(registryComponentName)
-		return nil, fmt.Errorf("generating hauler store tar: %w", err)
+		return nil, fmt.Errorf("writing registry script: %w", err)
 	}
 
-	haulerBinaryPath := fmt.Sprintf("hauler-%s", string(ctx.ImageDefinition.Image.Arch))
-	err = copyHaulerBinary(ctx, haulerBinaryPath)
+	registriesDir := filepath.Join(ctx.CombustionDir, "registry")
```
Comment (atanasdinov) [inline pkg/combustion/registry.go:49]: The dir name should be a constant in order for us to be able to easily reuse it.

File: pkg/combustion/registry.go [hunk 6]
```diff
@@ -30,50 +28,48 @@ var haulerManifest string
...
 		log.AuditComponentFailed(registryComponentName)
-		return nil, fmt.Errorf("copying hauler binary: %w", err)
+		return nil, fmt.Errorf("creating registry dir: %w", err)
 	}
 
-	registryScriptNameResult, err := writeRegistryScript(ctx)
-	if err != nil {
-		log.AuditComponentFailed(registryComponentName)
-		return nil, fmt.Errorf("writing registry script: %w", err)
+	if len(ctx.ImageDefinition.EmbeddedArtifactRegistry.ContainerImages) != 0 {
```
Comment (atanasdinov) [inline pkg/combustion/registry.go:56]: We're missing the Helm charts implementation.

File: pkg/combustion/registry.go [hunk 7]
```diff
@@ -30,50 +28,48 @@ var haulerManifest string
...
 	return []string{registryScriptNameResult}, nil
 }
 
-func writeHaulerManifest(ctx *image.Context) error {
+func writeHaulerManifest(ctx *image.Context, images []image.ContainerImage, charts []image.HelmChart, haulerManifestYamlName string) error {
 	haulerManifestYamlFile := filepath.Join(ctx.BuildDir, haulerManifestYamlName)
+	registryDef := image.EmbeddedArtifactRegistry{ContainerImages: images, HelmCharts: charts}
 
-	data, err := template.Parse(haulerManifestYamlName, haulerManifest, ctx.ImageDefinition.EmbeddedArtifactRegistry)
+	data, err := template.Parse(haulerManifestYamlName, haulerManifest, registryDef)
```
Comment (atanasdinov) [inline pkg/combustion/registry.go:72]: A unified manifest is not what we want if we're creating separate tarballs. This makes the logic too hard to follow as well as having empty manifest sections.

File: pkg/combustion/registry.go [hunk 8]
```diff
@@ -173,6 +169,28 @@ func createRegistryCommand(ctx *image.Context, commandName string, args []string
 	return cmd, logFile, nil
 }
 
-func IsEmbeddedArtifactRegistryEmpty(registry image.EmbeddedArtifactRegistry) bool {
-	return len(registry.HelmCharts) == 0 && len(registry.ContainerImages) == 0
+func IsEmbeddedArtifactRegistryAndKubernetesManifestsEmpty(ctx *image.Context) bool {
+	return len(ctx.ImageDefinition.EmbeddedArtifactRegistry.ContainerImages) == 0 && len(ctx.ImageDefinition.Kubernetes.HelmCharts) == 0
+}
+
+func writeHaulerManifestAndGenerateTar(ctx *image.Context, images []image.ContainerImage, charts []image.HelmChart, registryOrigin string) error {
```
Comment (atanasdinov) [inline pkg/combustion/registry.go:176]: On my previous note, once you start having to call functions `doThisAndThis`, they're no longer easy to understand, maintain and test. What we can do here is to split the flows and have `configureContainerImages` and `configureHelmCharts` functions or something similar. Each one would be working with its separate template and would be called based on its own condition e.g. 
```
if len(ctx.ImageDefinition.EmbeddedArtifactRegistry.ContainerImages) != 0 {
   configureContainerImages()
}
```

This will allow better testing and easier upgrades once we get to adding manifests in as well.

File: pkg/combustion/templates/13-embedded-registry.sh.tpl (hunk 2 above)
Comment (dbw7) [reply pkg/combustion/templates/13-embedded-registry.sh.tpl:7]: Sounds good 👍

File: pkg/combustion/registry.go (hunk 3 above)
Comment (dbw7) [reply pkg/combustion/registry.go:172]: Will do👍

**Comment by dbw7** [comment]
Closing this PR based on the implementation changing and being present in #123 .
